cdap-docs/developers-manual/source/building-blocks/workers.rst

```diff
@@ -0,0 +1,71 @@
+.. meta::
+    :author: Cask Data, Inc.
+    :copyright: Copyright © 2015 Cask Data, Inc.
+
+.. _workers:
+
+============================================
+Workers
+============================================
+
+Workers can be used to run background processes. Workers provide the ability to write data processing logic
+that doesn't fit into the other paradigms such as Flows and MapReduce or realtime and batch.
+
+You can add workers to your application by calling the ``addWorker`` method in the Application's
+``configure`` method::
+
+  public class AnalyticsApp extends AbstractApplication {
+    @Override
+    public void configure() {
+      setName("AnalyticsApp");
+      ...
+      addWorker(new ProcessWorker());
+      ...
+    }
+  }
+
+Workers have semantics similar to a Java thread and are run in a thread when CDAP is run in either In-Memory
+or Standalone Modes. In distributed mode, each instance of a worker runs in its own YARN container.
+Their instances may be updated via the :ref:`Command Line Interface <cli-available-commands>` or a :ref:`RESTful API <http-restful-api-lifecycle>`::
+
+  public class ProcessWorker extends AbstractWorker {
+
+    @Override
+    public void initialize(WorkerContext context) {
+      ...
+    }
+
+    @Override
+    public void run() {
+      ...
+    }
+
+    @Override
+    public void stop() {
+      ...
+    }
+
+    @Override
+    public void destroy() {
+      ...
+    }
+  }
+
+Workers can access and use ``Dataset``\s via a ``DatasetContext`` inside their ``run`` method::
+
+  @Override
+  public void run() {
```

Comment:
Might also want to show how you can do service discovery from a Worker. In the past we've had a lot of confusion regarding this because of lack of documentation.